src/java/org/apache/cassandra/service/accord/async/AsyncOperation.java

```diff
@@ -123,55 +131,114 @@ protected void setState(State state)
         this.state = state;
     }
 
-    /**
-     * callback for loader and writer
-     */
-    @Override
-    public void accept(Object o, Throwable throwable)
+    private void callback(Object o, Throwable throwable)
     {
         if (throwable != null)
         {
             logger.error(String.format("Operation %s failed", this), throwable);
             state = State.FAILED;
-            tryFailure(throwable);
+            fail(throwable);
```

Comment:
think you need to call `releaseResources` (well something like it), if you attempt to load 5 commands and 1 fails, then the whole operation fails and we don't release the 4 successful commands.

Another solution would be to change `org.apache.cassandra.service.accord.async.AsyncLoader#referenceAndDispatchReads` to add a cleanup on-error handler.  You would need to do this at the top level so a failure to load CommandsForKey releases Command as well.  Since failed loads still up the rc I think you can just release the keys without issue...  though may have to worry about what thread is doing the work to avoid race conditions

Reply:
We probably want to introduce some fuzz testing for this and the other related cleanup behaviours.

Reply:
I took a stab to start a fuzz test for this, and found issues with write path as well, see https://github.com/dcapwell/cassandra/tree/feedback/immutable-state

https://github.com/dcapwell/cassandra/commit/eef07eef25300434919e5fb118305c7137535e68

The bug was the following

1) the write actually fails, but we de-ref Node... so we now have a dirty write in-memory that was not durable
2) CommandsForKeys cache isn't cleaned up (test detects this and fails)

Reply:
Ok for the CommandsForKey cache thing, it seems to happen when there are failures and when there are no failures... the test doesn't show this as it expects the write to have failed (but we ignore write failures), so need to change to `awaitUninterruptibly(operation);` to check the references

Reply:
for #2 the issue looks to be that `org.apache.cassandra.service.accord.AccordKeyspace#loadCommandsForKey` returns `EMPTY`, and when its empty we don't lower ref count